pkg/loop/logger.go

```diff
@@ -52,6 +55,81 @@ func removeArg(args []interface{}, key string) ([]interface{}, string) {
 	return args, ""
 }
 
+// logMessage is the JSON payload that gets sent to Stderr from the plugin to the host
+type logMessage struct {
+	Message   string                 `json:"@message"`
+	Level     string                 `json:"@level"`
+	Timestamp time.Time              `json:"timestamp"`
+	ExtraArgs []*LogMessageExtraArgs `json:"extra_args"`
+}
+
+// LogMessageExtraArgs is a key value pair within the Output payload
+type LogMessageExtraArgs struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value"`
+}
+
+// flattenExtraArgs is used to flatten arguments of the log message
+func flattenExtraArgs(le *logMessage) []interface{} {
+	var result []interface{}
+	result = append(result, "level")
+	result = append(result, le.Level)
+	result = append(result, "timestamp")
+	result = append(result, le.Timestamp)
+	for _, kv := range le.ExtraArgs {
+		result = append(result, kv.Key)
+		result = append(result, kv.Value)
+	}
+
+	return result
+}
+
+// parseJSON handles parsing JSON output
+func parseJSON(input string) (*logMessage, error) {
+	var raw map[string]interface{}
+	entry := &logMessage{}
+
+	err := json.Unmarshal([]byte(input), &raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if v, ok := raw["@message"]; ok {
+		entry.Message = v.(string)
+		delete(raw, "@message")
+	}
+
+	if v, ok := raw["@level"]; ok {
+		entry.Level = v.(string)
+		delete(raw, "@level")
+	}
+
+	for k, v := range raw {
+		entry.ExtraArgs = append(entry.ExtraArgs, &LogMessageExtraArgs{
+			Key:   k,
+			Value: v,
+		})
+	}
+
+	return entry, nil
+}
+
+// logDebug will parse msg and figure out if it's a panic, fatal or critical log message, this is done here because the hashicorp plugin will push any
+// unrecognizable message from stderr as a debug statement
+func logDebug(msg string, l logger.Logger, args ...interface{}) {
+	if strings.HasPrefix(msg, "panic:") {
+		l.Errorw(msg, args...)
+	} else if log, err := parseJSON(msg); err == nil {
+		if log.Level == "dpanic" || log.Level == "fatal" || log.Level == "critical" {
+			l.Errorw(log.Message, flattenExtraArgs(log)...)
```

Comment:
Should we add a prefix here? Otherwise they just look like regular error level, right?

Reply:
Oh wait, we can just forward to the correct method, right?
```suggestion
			l.Fatalw(log.Message, flattenExtraArgs(log)...)
```

Reply:
which process would be exit'd? the loop or the core node? seems correct to exit the loop and definitely wrong to exit the core node

Reply:
Wouldn't be helpful to at least state when it is a panic, a fatal or a critical log level in the log when we forward the message?

I feel it would be good to understand the severity and not encapsulate/group all 3 error levels in a single error, if that makes sense.

Reply:
@krehermann it's going to exit the core node
@vyzaldysanchez we can try prefixing it, it will come out a bit ugly like `[ERROR] [CRIT] some message` but if that's fine we can go with it

Reply:
Oh right, no fatal. But what about critical?

And I think we could escalate the panic logs to ciritical level too.

Reply:
Added prefixes and escalated the logs